wagtail/migrations/0084_fix_log_entry_json_timestamps.py

```diff
@@ -0,0 +1,137 @@
+
+
+def migrate_logs_with_created_only(model, converter):
+    for item in (
+        model.objects.filter(
+            action__in=["wagtail.revert", "wagtail.rename", "wagtail.publish"]
+        )
+        .only("data")
+        .iterator()
+    ):
+        try:
+            # If a previous_revision was available, the data contains "revision" with
+            # its created date.
+            # Also, there are "wagtail.publish" logs, which don't set data at all.
+            created = item.data["revision"]["created"]
+            # "created" is set to the previous revision's created_at, which is set
+            # to UTC by django.
+            item.data["revision"]["created"] = converter(created)
+        except ValueError:
+            # TODO TBD: log error?
+            continue
+        except KeyError:
+            continue
+        else:
+            item.save(update_fields=["data"])
+
+
+def migrate_schedule_logs(model, converter):
+    for item in (
+        model.objects.filter(
+            action__in=["wagtail.publish.schedule", "wagtail.schedule.cancel"]
+        )
+        .only("data")
+        .iterator()
+    ):
+        created = item.data["revision"]["created"]
+        # May be unset for "wagtail.schedule.cancel"-logs.
+        go_live_at = item.data["revision"].get("go_live_at")
+        try:
+            # "created" is set to timezone.now() for new revisions ("wagtail.publish.schedule")
+            # and to self.created_at for "wagtail.schedule.cancel", which is set to UTC
+            # by django.
+            item.data["revision"]["created"] = converter(created)
+
+            if go_live_at:
+                # The go_live_at date is set to the revision object's "go_live_at".
+                # The revision's object is created by deserializing the json data (see wagtail.models.Revision.as_object()),
+                # and this process converts all datetime objects to the local timestamp (see https://github.com/wagtail/django-modelcluster/blob/8666f16eaf23ca98afc160b0a4729864411c0563/modelcluster/models.py#L109-L115).
+                # That's the reason, why this date is the only date, which is not stored in the log's JSON as UTC, but in the default timezone.
+                item.data["revision"]["go_live_at"] = converter(
+                    go_live_at, tz=timezone.get_default_timezone()
+                )
+        except ValueError:
+            # TODO TBD: log error?
+            continue
+        else:
+            item.save(update_fields=["data"])
+
+
+def migrate_custom_to_iso_format(apps, schema_editor):
+    ModelLogEntry = apps.get_model("wagtailcore.ModelLogEntry")
+    PageLogEntry = apps.get_model("wagtailcore.PageLogEntry")
+
+    migrate_logs_with_created_only(ModelLogEntry, legacy_to_iso_format)
+    migrate_logs_with_created_only(PageLogEntry, legacy_to_iso_format)
+
+    migrate_schedule_logs(ModelLogEntry, legacy_to_iso_format)
+    migrate_schedule_logs(PageLogEntry, legacy_to_iso_format)
```

Comment:
Note that we support [custom `FooLogEntry` classes](https://docs.wagtail.org/en/stable/extending/audit_log.html#log-models). I'm not sure if we should try to migrate those logs as well or leave it to the developers (at least with a mention in the upgrade considerations, linking to this migration file). I don't know if it's even possible to do so though, as the log model registry uses real models and I don't know if we can get the fake model from there to use in migrations.

Reply:
Yes, loading those models without loading the real models looks impossible to me.
I think we can only mention this in the upgrade notes, maybe together with the undocumented way of using different action names mentioned in the todo-notes above.